x-pack/plugins/observability/public/components/shared/exploratory_view/hooks/use_add_to_case.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useMemo, useState } from 'react';
+import { isEmpty } from 'lodash';
+import { i18n } from '@kbn/i18n';
+import { HttpSetup, MountPoint } from 'kibana/public';
+import { useKibana } from '../../../../utils/kibana_react';
+import { Case, SubCase } from '../../../../../../cases/common';
+import { TypedLensByValueInput } from '../../../../../../lens/public';
+import { AddToCaseProps } from '../header/add_to_case_action';
+
+const appendSearch = (search?: string) =>
+  isEmpty(search) ? '' : `${search?.startsWith('?') ? search : `?${search}`}`;
+
+const getCreateCaseUrl = (search?: string | null) =>
+  `/cases/create${appendSearch(search ?? undefined)}`;
+
+const appId = 'observability';
+
+async function addToCase(
+  http: HttpSetup,
+  theCase: Case | SubCase,
+  attributes: TypedLensByValueInput['attributes'],
+  timeRange: { from: string; to: string }
+) {
+  const apiPath = `/api/cases/${theCase?.id}/comments`;
+
+  const vizPayload = {
+    attributes,
+    timeRange,
+  };
+
+  const payload = {
+    comment: `!{lens${JSON.stringify(vizPayload)}}`,
+    type: 'user',
+    owner: 'observability',
+  };
+
+  return http.post(apiPath, { body: JSON.stringify(payload) });
+}
+
+export const useAddToCase = ({
+  lensAttributes,
+  getToastText,
+  timeRange,
+}: AddToCaseProps & { getToastText: (thaCase: Case | SubCase) => MountPoint<HTMLElement> }) => {
+  const [isSaving, setIsSaving] = useState(false);
+  const [isCasesOpen, setIsCasesOpen] = useState(false);
+
+  const {
+    http,
+    application: { navigateToApp, getUrlForApp },
+    notifications: { toasts },
+  } = useKibana().services;
+
+  const createCaseUrl = useMemo(() => getUrlForApp(appId) + getCreateCaseUrl(), [getUrlForApp]);
+
+  const goToCreateCase = useCallback(
+    async (ev) => {
+      ev.preventDefault();
+      return navigateToApp(appId, {
+        path: getCreateCaseUrl(),
+      });
+    },
+    [navigateToApp]
+  );
+
+  const onCaseClicked = useCallback(
```

Comment:
You can define this as a async function rather than using `.then` syntax.

Reply:
i actually wanted to add error handling here. i find erro handling in this pattern much more readable as compared to try/catch approach with await.